sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -421,48 +421,515 @@ case class Cast(child: Expression, dataType: DataType) extends UnaryExpression w
+    val toValuePrim = ctx.freshName("tvp")
+    val toValueNull = ctx.freshName("tvn")
+    val result = ctx.freshName("result")
+
+
+    CodeHolder((c, evPrim, evNull) =>
+      s"""
+        final $hashMapClass $result = new $hashMapClass();
+        scala.collection.Iterator iter = $c.iterator();
+        while (iter.hasNext()) {
+          scala.Tuple2 kv = (scala.Tuple2) iter.next();
+          boolean $fromKeyNull = false;
+          ${ctx.boxedType(from.keyType)} $fromKeyPrim =
+            (${ctx.boxedType(from.keyType)}) kv._1();
+          ${keyCodeHolder.code(ctx, fromKeyPrim,
+            fromKeyNull, toKeyPrim, toKeyNull, to.keyType)}
+
+          boolean $fromValueNull = kv._2() == null;
+          if ($fromValueNull) {
+            $result.put($toKeyPrim, null);
+          } else {
+            ${ctx.boxedType(from.valueType)} $fromValuePrim =
+              (${ctx.boxedType(from.valueType)}) kv._2();
+            ${valueCodeHolder.code(ctx, fromValuePrim,
+              fromValueNull, toValuePrim, toValueNull, to.valueType)}
+            if ($toValueNull) {
+              $result.put($toKeyPrim, null);
+            } else {
+              $result.put($toKeyPrim, $toValuePrim);
+            }
+          }
+        }
+        $evPrim = $result;
+      """)
+  }
 
-      case (_: NumericType, dt: NumericType) =>
-        defineCodeGen(ctx, ev, c => s"(${ctx.javaType(dt)})($c)")
+  private[this] def castStructCode(
+      from: StructType, to: StructType, ctx: CodeGenContext): CodeHolder = {
 
-      case other =>
-        super.genCode(ctx, ev)
+    val fieldsCodeHolder = from.fields.zip(to.fields).map {
+      case (fromField, toField) => getCodeHolder(fromField.dataType, toField.dataType, ctx)
     }
+    val rowClass = "org.apache.spark.sql.catalyst.expressions.GenericMutableRow"
+    val result = ctx.freshName("result")
+    val tmpRow = ctx.freshName("tmpRow")
+
+    val fieldsEvalCode = fieldsCodeHolder.zipWithIndex.map { case (holder, i) => {
+      val fromFieldPrim = ctx.freshName("ffp")
+      val fromFieldNull = ctx.freshName("ffn")
+      val toFieldPrim = ctx.freshName("tfp")
+      val toFieldNull = ctx.freshName("tfn")
+      val fromType = ctx.boxedType(from.fields(i).dataType)
+      s"""
+        boolean $fromFieldNull = $tmpRow.isNullAt($i);
+        if ($fromFieldNull) {
+          $result.update($i, null);
+        } else {
+          $fromType $fromFieldPrim = ($fromType) $tmpRow.apply($i);
+          ${holder.code(ctx, fromFieldPrim,
+            fromFieldNull, toFieldPrim, toFieldNull, to.fields(i).dataType)}
+          if ($toFieldNull) {
+            $result.update($i, null);
+          } else {
+            $result.update($i, $toFieldPrim);
+          }
+        }
+       """
+      }
+    }.mkString("\n")
+
+    CodeHolder((c, evPrim, evNull) =>
+      s"""
+        final $rowClass $result = new $rowClass(${fieldsCodeHolder.size});
+        final InternalRow $tmpRow = $c;
+        $fieldsEvalCode
```

Comment:
I'm unfolding the field's evaluation here since it's only possible to determine each field's calculation code here.